crates/wasmtime/src/trap.rs

```diff
@@ -237,3 +317,44 @@ impl From<Box<dyn std::error::Error + Send + Sync>> for Trap {
         }
     }
 }
+
+#[test]
+fn heap_out_of_bounds_trap() {
+    let store = crate::Store::default();
+    let module = crate::Module::new(
+        store.engine(),
+        r#"
+            (module
+              (memory 0)
+              (func $start (drop (i32.load (i32.const 1000000))))
+              (start $start)
+            )
+         "#,
+    )
+    .unwrap();
+
+    let err = match crate::Instance::new(&store, &module, &[]) {
+        Ok(_) => unreachable!(),
+        Err(e) => e,
+    };
+    let trap = err.downcast_ref::<Trap>().unwrap();
+    assert_eq!(trap.trap_code(), Some(TrapCode::MemoryOutOfBounds));
+
+    let module = crate::Module::new(
+        store.engine(),
+        r#"
+            (module
+              (memory 0)
+              (func $start (drop (i32.load memory.size)))
+              (start $start)
+            )
+         "#,
+    )
+    .unwrap();
+    let err = match crate::Instance::new(&store, &module, &[]) {
+        Ok(_) => unreachable!(),
+        Err(e) => e,
+    };
+    let trap = err.downcast_ref::<Trap>().unwrap();
+    assert_eq!(trap.trap_code(), Some(TrapCode::MemoryOutOfBounds));
+}
```

Comment:
Could this get moved to `tests/all/*.rs`? That's were we store most other tests.

Reply:
Also, while you're at it, mind simplifying this a bit with a helper function?

```rust
fn assert_trap_code(wat: &str, code: ir::TrapCode)
```

or something like that?